clients/src/main/java/org/apache/kafka/common/security/kerberos/Login.java

```diff
@@ -0,0 +1,410 @@
+                                    // sleep for 10 seconds.
+                                    try {
+                                        Thread.sleep(10 * 1000);
+                                    } catch (InterruptedException e) {
+                                        log.error("Interrupted during login retry after LoginException:", le);
+                                        throw le;
+                                    }
+                                } else {
+                                    log.error("Could not refresh TGT for principal: " + principal + ".", le);
+                                }
+                            }
+                        }
+                    } catch (LoginException le) {
+                        log.error("Failed to refresh TGT: refresh thread exiting now.", le);
+                        break;
+                    }
+                }
+            }
+        });
+        t.setDaemon(true);
+    }
+
+    public void startThreadIfNeeded() {
+        // thread object 't' will be null if a refresh thread is not needed.
+        if (t != null) {
+            t.start();
+        }
+    }
+
+    public void shutdown() {
+        if ((t != null) && (t.isAlive())) {
+            t.interrupt();
+            try {
+                t.join();
+            } catch (InterruptedException e) {
+                log.warn("error while waiting for Login thread to shutdown: " + e);
+            }
+        }
+    }
+
+    public Subject subject() {
+        return subject;
+    }
+
+    public String loginContextName() {
+        return loginContextName;
+    }
+
+    private synchronized LoginContext login(final String loginContextName) throws LoginException {
+        if (loginContextName == null) {
+            throw new LoginException("loginContext name (JAAS file section header) was null. " +
+                    "Please check your java.security.login.auth.config (=" +
+                    System.getProperty("java.security.login.auth.config") +
+                    ") and your " + JaasUtils.LOGIN_CONTEXT_SERVER + "(=" +
+                    System.getProperty(JaasUtils.LOGIN_CONTEXT_CLIENT, "Client") + ")");
+        }
+
+        if (System.getProperty("java.security.auth.login.config") == null) {
+            throw new IllegalArgumentException("You must pass java.security.auth.login.config in secure mode.");
+        }
+
+        File configFile = new File(System.getProperty("java.security.auth.login.config"));
+        Configuration loginConf = null;
+        try {
+            loginConf = Configuration.getInstance("JavaLoginConfig", new URIParameter(configFile.toURI()));
+            Configuration.setConfiguration(loginConf);
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException(e);
+        }
+
+        LoginContext loginContext = new LoginContext(loginContextName, callbackHandler);
+        loginContext.login();
+        log.info("successfully logged in.");
```

Comment:
There was this code in the constructor before the `login` call:

`this.lastLogin = time.currentElapsedTime() - this.minTimeBeforeRelogin;`

I've changed it to:

`this.lastLogin = time.currentElapsedTime()`

`minTimeBeforeRelogin` is only relevant for the relogin case. However, we are still setting the `lastLogin` time before we actually execute `loginContext.login` (in both the first and subsequent logins). Do you think we should be updating that value after the `loginContext.login` call?